src/backend/replication/logical/message.c

```diff
@@ -93,3 +96,62 @@ logicalmsg_redo(XLogReaderState *record)
 
 	/* This is only interesting for logical decoding, see decode.c. */
 }
+
+/*
+ * NEON: persist file in WAL to save it in persistent storage.
+ * If fd < 0, then remote entry from page server.
+ */
+void
+wallog_file_descriptor(char const* path, int fd, uint64_t limit)
+{
+	char	prefix[MAXPGPATH];
+
+	/* Do not wallog AUX file at replica */
+	if (!XLogInsertAllowed())
+		return;
+
+	snprintf(prefix, sizeof(prefix), "neon-file:%s", path);
+	if (fd < 0)
+	{
+		elog(DEBUG1, "neon: deleting contents of file %s", path);
+		/* unlink file */
+		LogLogicalMessage(prefix, NULL, 0, false, true);
+	}
+	else
+	{
+		off_t size = lseek(fd, 0, SEEK_END);
+		elog(DEBUG1, "neon: writing contents of file %s, size %ld", path, (long)size);
+		if (size < 0)
+			elog(ERROR, "Failed to get size of file %s: %m", path);
+		if ((uint64_t)size > limit)
+		{
+			elog(WARNING, "Size of file %s %ld is larger than limit %ld", path, (long)size, (long)limit);
+		}
+		else
+		{
+			char* buf = palloc((size_t)size);
+			lseek(fd, 0, SEEK_SET);
+			if (read(fd, buf, (size_t)size) != size)
+				elog(ERROR, "Failed to read file %s: %m", path);
```

Comment:
Needs to handle short reads.

Rather than call lseek(), you could use pread().

Reply:
Fixed